File: BOJ/C++/done/15711.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;
using ll = long long;
int t;
ll a, b;
vector<int> prime;
vector<bool> isPrime(1.5e6, true);

bool check(ll a, ll b)
{
    ll sum = a + b;
    if (sum == 2)
        return false;
    if (sum % 2 == 0)
        return true;
    if ((sum - 2) < 1.5e6)
        return isPrime[sum - 2];
    for (int i = 0; i < prime.size(); i++)
    {
        if ((sum - 2) % prime[i] == 0)
            return false;
    }
    return true;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    string ans = "";
    isPrime[0] = isPrime[1] = false;

    for (int i = 2; i * i < isPrime.size(); i++)
    {
        for (int j = i + i; j < isPrime.size(); j += i)
            isPrime[j] = false;
    }

    for (int i = 2; i < isPrime.size(); i++)
    {
        if (isPrime[i])
            prime.push_back(i);
    }

    cin >> t;
    while (t--)
    {
        cin >> a >> b;
        if (check(a, b))
            ans += "YES\n";
        else
            ans += "NO\n";
    }

    cout << ans;
    return 0;
}

File: BOJ/JS/done/15711.js
const fs = require('fs');
const input = fs.readFileSync('./input.txt').toString().trim().split('\n');
const t = +input[0];
const ropes = input.slice(1);
const ans = [];
const isPrime = Array(1.5e6).fill(true);
const prime = [];
isPrime[0] = isPrime[1] = false;

for (let i = 2; i * i < isPrime.length; i++) {
  for (let j = i + i; j < isPrime.length; j += i) {
    isPrime[j] = false;
  }
}
isPrime.forEach((b, idx) => {
  if (b) prime.push(idx);
});

function check(a, b) {
  const sum = a + b;
  if (a + b === 2) return false;
  if ((a + b) % 2 === 0) return true;
  if (sum - 2 < isPrime.length) return isPrime[sum - 2];
  for (let i = 0; i < prime.length; i++) {
    if ((sum - 2) % prime[i] === 0) return false;
  }
  return true;
}

ropes.forEach((v) => {
  const [a, b] = v.split(' ').map(Number);
  check(a, b) ? ans.push('YES') : ans.push('NO');
});

console.log(ans.join('\n'));
